src/awesome/drawable.rs

```diff
@@ -1,24 +1,35 @@
 //! TODO Fill in
 
+use cairo::{Format, ImageSurface};
+use glib::translate::ToGlibPtr;
+use rustwlc::Geometry;
 use std::fmt::{self, Display, Formatter};
 use std::default::Default;
 use std::rc::Rc;
-use rlua::{self, Table, Lua, UserData, ToLua, Value};
+use rlua::{self, Table, Lua, UserData, ToLua, Value, LightUserData};
 use super::object::{Object, Objectable};
-use super::class::{self, Class, ClassBuilder};
+use super::class::{self, Class};
+use super::property::Property;
 
 #[derive(Clone, Debug)]
 pub struct DrawableState {
-    // TODO Fill in
-    dummy: i32
+    // TODO Pixmap?
+    pub surface: Option<ImageSurface>,
+    geo: Geometry,
+    refreshed: bool,
+    // TODO Refresh callback
```

Comment:
You are using an ImageSurface which allocates its own memory to draw to; awesome uses a XcbSurface drawing to the pixmap. Thus, you don't need a pixmap member here, it's just the "target" for drawing in awesome.

The refresh callback is just an internal implementation detail. For drawins it copies from the drawable's pixmap to the screen and for titlebars it does basically the same. However, for your case it should be enough to just always copy the contents of the surface to the screen on each frame and basically ignore the refresh function that Lua calls (it is always valid to use the contents of the surface, the refresh callback from Lua is just a "I changed something"-signal to force a repaint; this can of course be optimized later).

Reply:
Thanks, I had a feeling I didn't need a pixmap but I hadn't looked into whether or not I needed it.

Ok so I don't even really need to add it as a field, I can just have a refresh function that checks the bool and only draws if it's true then. That will simplify this.